### ISSUE TITLE
Make ownedNFTs `access(self)` instead of `pub`

### ISSUE THREAD

**Issue by rheaplex** [onflow/flow-nft#45]
Should these member variables:

https://github.com/onflow/flow-nft/blob/master/contracts/NonFungibleToken.cdc#L116

https://github.com/onflow/flow-nft/blob/master/contracts/ExampleNFT.cdc#L29

```
        pub var ownedNFTs: @{UInt64: NonFungibleToken.NFT}
```

instead be declared like this:

```
        access(self) let ownedNFTs: @{UInt64: NonFungibleToken.NFT}
```

As `pub` access allows the ownedNFTs dictionary to be manipulated without going through the deposit/withdraw functions, and `var` allows it to be replaced wholesale.

If this is desired behaviour, it is worth documenting the implications of it in the code.

**Comment by bluesign** [comment]
I think current one is the correct approach. It is a bit balance between user-dapp power.

**Comment by rheaplex** [comment]
If the field is for modification by the user, it should probably be called `userMetadata` or `notes`.

The current implementation has enough subtleties (the field is user settable, and this means that the Collection API can be worked around) that I do feel that its presence in sample code should either be carefully explained or replaced with something that better embodies the "metadata" patterns that we see in e.g. TopShot.

I lean very strongly towards the latter, as people are going to copy-and-paste this code for projects and we can make this more robust for them with maybe three lines of code changed.

**Comment by bluesign** [comment]
This is not metadata, this is where NFTs are stored on a collection. 

As far as I understand, "Collection API can be worked around" is 'by design'.

When a user buys an NFT, it should be forever theirs. So if they put it in a collection, they still have access to underlying storage (**ownedNFTs**). They can get the token anytime they want. Making that field `access(self) `is taking that power from the user.

Collection withdraw and deposit functions operating on that storage, like helper functions.

**Comment by rheaplex** [comment]
Yes, sorry, I responded on the wrong ticket. 😞 I've deleted that.

I agree with the case you make here. Thank you. 🙂

I think it is worth documenting the implications of the design; that the owner can always access their tokens and so schemes of the kind you identify can always be worked around. And further noting that the deposit/withdraw etc. member functions are for exposing the ability to manipulate the collection granularly through *capabilities* rather than representing an object-oriented-style interface to encapsulated state.

**Comment by joshuahannan** [comment]
I think more discussion is needed for this issue. I think it is worthwhile considering making the field private so projects can have the ability to include special deposit or withdraw functions if they wish, such as for paying royalties or other such things.

**Comment by bluesign** [comment]
@joshuahannan I don't think this one is easy, this would require `deposit` and `withdraw` to return collection, also then same thing as NFT, user's can trade collections instead of NFTs :) 

There is no way to make a forced royalty system in cadence philosophy.

**Comment by rheaplex** [comment]
> @joshuahannan I don't think this one is easy, this would require `deposit` and `withdraw` to return collection, also then same thing as NFT, user's can trade collections instead of NFTs :)

Here's a quick sketch of an entirely updated `Collection` type that avoids this issue.

I'm not proposing or recommending this approach, just demonstrating its possibility.

```
pub resource CollectionDesignThatPreventsThis {
  access(self) _tokens: @{UInt64: NFT}

  access(contract) pub fun _withdraw(tokenID: UInt64): @NFT {
    return self._tokens[tokenID]
  }

  access(contract) pub fun _deposit(token: @NFT} {
    self._tokens[token.id] = token
  }

  pub fun transfer(collection: Collection, tokenID: UInt64) {
    self._deposit(collection._withdraw(tokenID)
  }
}
```

Mint functions must use the internal functions to deposit the NFT directly in order to keep it captured.

**Comment by bluesign** [comment]
Yeah but this one is just encapsulates the NFT inside collection ( actually when I suggested withdrawal and deposit returning Collection, I suggested this pattern ) User doesn't have access to NFT.

But now they can still move NFT to emptyCollection and trade collection instead to skip paying royalties. Point is technically not possible to force royalties in Flow. 

I don't think user's access to NFT should be limited, but I think this is an another discussion of cadence rules.

**Comment by joshuahannan** [comment]
@rheaplex I don't think your suggestion works because like @bluesign said, people need to be able to access their NFTs for composability sake. I need to be able to withdraw my NFT from my collection in case there is a third party project that has a special type that stores an NFT or something.

@bluesign I agree that users need to have access, but I think that it is probably up to the project to decide what the rules are for deposit and withdraw. If a user decides to participate in that project, they are doing so while agreeing to the rules of the smart contract. If they want a project where they have full access to their NFTs, they should go find a different one.

And I understand that many users will be able to find ways to get around paying royalties, but the point isn't to ensure that it is impossible to avoid royalties, the point is to provide a standard way to do that, which most people will agree to do because it is easiest and safest.

**Comment by joshuahannan** [comment]
Closing this because it is included in #48
